File: ApiGateway/internal/handlers/health.go
package handlers

import (
	"encoding/json"
	"net/http"
	"time"
)

var (
	startTime      = time.Now().UTC() // Almacenamos el tiempo de inicio del servicio
	serviceVersion = "1.0.0"          // Definimos la versión del servicio
)

type Check struct {
	Data   map[string]string `json:"data"`
	Name   string            `json:"name"`
	Status string            `json:"status"`
}

type HealthResponse struct {
	Status  string  `json:"status"`
	Checks  []Check `json:"checks"`
	Details struct {
		Version string `json:"version"`
		Uptime  string `json:"uptime"`
	} `json:"details"`
}

// Liveness check - /health/live
func LivenessCheckHandler(w http.ResponseWriter, r *http.Request) {
	uptime := time.Since(startTime).String()
	checkData := map[string]string{
		"from":   time.Now().UTC().Format(time.RFC3339),
		"status": "ALIVE",
	}

	response := HealthResponse{
		Status: "UP",
		Checks: []Check{
			{
				Data:   checkData,
				Name:   "Liveness check",
				Status: "UP",
			},
		},
	}

	response.Details.Version = serviceVersion
	response.Details.Uptime = uptime

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

// Readiness check - /health/ready
func ReadinessCheckHandler(w http.ResponseWriter, r *http.Request) {
	isDBConnected := true // Simulamos la conexión a la base de datos
	status := "UP"
	if !isDBConnected {
		status = "DOWN"
	}

	uptime := time.Since(startTime).String()
	checkData := map[string]string{
		"from":   time.Now().UTC().Format(time.RFC3339),
		"status": "READY",
	}

	response := HealthResponse{
		Status: status,
		Checks: []Check{
			{
				Data:   checkData,
				Name:   "Readiness check",
				Status: status,
			},
		},
	}

	response.Details.Version = serviceVersion
	response.Details.Uptime = uptime

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

// General health status - /health
func GeneralHealthHandler(w http.ResponseWriter, r *http.Request) {
	uptime := time.Since(startTime).String()
	currentTime := time.Now().UTC().Format(time.RFC3339)

	response := HealthResponse{
		Status: "UP",
		Checks: []Check{
			{
				Data:   map[string]string{"from": currentTime, "status": "READY"},
				Name:   "Readiness check",
				Status: "UP",
			},
			{
				Data:   map[string]string{"from": currentTime, "status": "ALIVE"},
				Name:   "Liveness check",
				Status: "UP",
			},
		},
	}

	response.Details.Version = serviceVersion
	response.Details.Uptime = uptime

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

File: ApiPerfiles/app/routers/health.py
from fastapi import APIRouter
from datetime import datetime, timedelta, timezone

router = APIRouter()

# Guardamos el tiempo de inicio del servicio
start_time = datetime.now(timezone.utc)  # Usamos timezone-aware datetime
service_version = "1.0.0"  # Definimos la versión del servicio

# Liveness check - /health/live
@router.get("/health/live", response_model=dict)
async def liveness_check():
    uptime = datetime.now(timezone.utc) - start_time
    response = {
        "status": "UP",
        "checks": [
            {
                "data": {
                    "from": datetime.now(timezone.utc).isoformat(),
                    "status": "ALIVE"
                },
                "name": "Liveness check",
                "status": "UP"
            }
        ],
        "details": {
            "version": service_version,
            "uptime": str(uptime)  # Convertimos timedelta a string legible
        }
    }
    return response

# Readiness check - /health/ready
@router.get("/health/ready", response_model=dict)
async def readiness_check():
    is_db_connected = True  # Simulamos conexión a base de datos
    status = "UP" if is_db_connected else "DOWN"
    
    uptime = datetime.now(timezone.utc) - start_time
    response = {
        "status": status,
        "checks": [
            {
                "data": {
                    "from": datetime.now(timezone.utc).isoformat(),
                    "status": "READY"
                },
                "name": "Readiness check",
                "status": status
            }
        ],
        "details": {
            "version": service_version,
            "uptime": str(uptime)
        }
    }
    return response

# General health status - /health
@router.get("/health", response_model=dict)
async def general_health():
    uptime = datetime.now(timezone.utc) - start_time
    response = {
        "status": "UP",
        "checks": [
            {
                "data": {
                    "from": datetime.now(timezone.utc).isoformat(),
                    "status": "READY"
                },
                "name": "Readiness check",
                "status": "UP"
            },
            {
                "data": {
                    "from": datetime.now(timezone.utc).isoformat(),
                    "status": "ALIVE"
                },
                "name": "Liveness check",
                "status": "UP"
            }
        ],
        "details": {
            "version": service_version,
            "uptime": str(uptime)
        }
    }
    return response
